kolibri/core/content/utils/content_request.py

```diff
@@ -0,0 +1,525 @@
+    return (
+        ContentDownloadRequest.objects.filter(status__in=INCOMPLETE_STATUSES)
+        .order_by("requested_at")
+        .annotate(
+            has_metadata=Exists(
+                ContentNode.objects.filter(pk=OuterRef("contentnode_id"))
+            ),
+            total_size=_total_size_annotation(),
+        )
+    )
+
+
+class InsufficientStorage(Exception):
+    """
+    Dedicated exception with which we can halt content request processing for insufficient storage
+    """
+
+    pass
+
+
+class NoPeerAvailable(Exception):
+    """
+    Dedicated exception with which we can halt content request processing when we don't have a peer
+    """
+
+    pass
+
+
+def process_content_requests():
+    """
+    Wrapper around the processing of content requests to capture errors
+    """
+    incomplete_downloads = _incomplete_downloads_queryset()
+
+    # first, process the metadata import for any incomplete downloads without metadata
+    incomplete_downloads_without_metadata = incomplete_downloads.filter(
+        has_metadata=False
+    )
+    if incomplete_downloads_without_metadata.exists():
+        _process_metadata_import(incomplete_downloads_without_metadata)
+
+    try:
+        _process_content_requests(incomplete_downloads)
+    except InsufficientStorage as e:
+        logger.warning(str(e))
+        LearnerDeviceStatus.save_statuses(DeviceStatus.InsufficientStorage)
+    except NoPeerAvailable as e:
+        logger.warning(str(e))
+
+
+def _get_import_metadata(client, contentnode_id):
+    """
+    :type client: NetworkClient
+    :type contentnode_id: str
+    :rtype: None|dict
+    """
+    url_path = reverse_path(
+        "kolibri:core:importmetadata-detail", kwargs={"pk": contentnode_id}
+    )
+    try:
+        response = client.get(url_path)
+        return response.json()
+    except NetworkLocationResponseFailure as e:
+        # 400 level errors, like 404, are ignored
+        if e.response and 400 <= e.response.status_code < 500:
+            logger.debug(
+                "Metadata request failure: GET {} {}".format(
+                    url_path, e.response.status_code
+                )
+            )
+            return None
+        raise e
+
+
+def _import_metadata(client, contentnode_ids):
+    """
+    :type client: NetworkClient
+    :param contentnode_ids: a values_list QuerySet of content node ids or list of them
+    :type contentnode_ids: QuerySet or list
+    """
+    total_count = (
+        contentnode_ids.count()
+        if isinstance(contentnode_ids, QuerySet)
+        else len(contentnode_ids)
+    )
+    processed_count = 0
+    logger.info("Importing content metadata for {} nodes".format(total_count))
+    for contentnode_id in contentnode_ids:
+        import_metadata = _get_import_metadata(client, contentnode_id)
+        # if the request 404'd, then we wouldn't have this data
+        if import_metadata:
+            processed_count += 1
+            with db_lock_sqlite_only():
```

Comment:
If we remove this context manager, does the error still happen?

Wondering if we should have some thread local property for `db_lock_sqlite_only` that prevents it trying to make a lock if it's already inside an existent lock?

Reply:
I'll need to test this again. With fresh eyes, I'm guessing this was still blocking the post-sync import of metadata for the same reason the sync caused it.